src/ui/PreviewModal.tsx

```diff
@@ -56,11 +58,29 @@ const PreviewModal: FC<Props & WithTranslation> = ({
               <InformationDisplayContainer preview monitor={view} />
             ) : (
               <>
-                {stationIds.length && showPlatformsOrTracks ? (
+                {(stations.length || stops.length) && showPlatformsOrTracks ? (
                   <TrainDataFetcher
                     monitor={view}
-                    stationIds={stationIds}
+                    stations={stations}
+                    stops={stops}
                     preview
+                    fetchOnlyHsl={
+                      instance === 'hsl'
+                        ? true
+                        : stations
+                            .concat(stops)
+                            .every(x => x.gtfsId.startsWith('HSL'))
+                    }
+                    fetchAlsoHsl={
+                      instance === 'hsl'
+                        ? false
+                        : !stations
+                            .concat(stops)
+                            .every(x => x.gtfsId.startsWith('HSL')) &&
+                          stations
+                            .concat(stops)
+                            .some(x => x.gtfsId.startsWith('HSL'))
+                    }
                   />
```

Comment:
This is bit unclear to me. Couldn't you just filter one array depending on the instance, either here or TDF? This seems overly complicated at first sight.

Reply:
- fetchOnlyHSL:
  * When instance is HSL or every stations/stop starts with HSL when instance is matka
-fetchAlsoHSL
  * When instance is not HSL but some stations/stops starts with HSL